docs/snapshotting/versioning.md

```diff
@@ -0,0 +1,98 @@
+
+This block diagram illustrates the concept:
+
+![Versionize](
+../images/versionize.png?raw=true
+"Versionize layers")
+
+## Bincode encoding
+During research and prototyping we have considered multiple storage formats. The criteria used for comparing these were performance, size, Rust support, specification, versioning support, community popularity and tooling. Performance, size and Rust support are hard requirements while all others will can be the subject of trade-offs. 
+
+More info about this comparison can be found [here](https://github.com/firecracker-microvm/firecracker/blob/9d427b33d989c3225d874210f6c2849465941dc0/docs/snapshotting/design.md).
+
+Key benefits of using `bincode` versus everything else are:
+- Minimal snapshot size overhead,
+- Minimal CPU overhead,
+- Simple implementation.
+
+## Snapshot compatibility
+### Host kernel
+
+The minimum kernel version required by Firecracker snapshots is 4.14. Snapshots can be saved and restored on different kernel versions without issues.
+
+### Device model
+The microVM state file links some resources that are external to the snapshot:
+- tap devices by device name,
+- block devices by block file path.
+
+To successfully restore a microVM one should check that:
+- tap devices are available, their names match their original names since these are the values saved in the microVM state file, and they are accessible to the Firecracker process where the microVM is being restored,
+- block devices are set up at their original relative or absolute paths with the proper permissions, as the Firecracker process with the restored microVM will attempt to access them exactly as they were accessed in the original Firecracker process.
+
+### CPU model
+
+Firecracker microVMs can run on any Intel x86_64 CPU models that support the hardware virtualizations extensions. Snapshots, however, are not compatible across all Intel x86_64 CPU models. They are only compatible if the CPU features exposed to the guest are an invariant when saving and restoring the snapshot.
+The trivial scenario is creating and restoring snapshots on hosts that have the same CPU model. 
+To relax this limitation, Firecracker provides an API to select a CPU template. Firecracker CPU templates work by using CPUID masking that restricts the exposed features to a common denominator of multiple CPU models.
+
+
+## Implementation
+
+To enable cross version snapshot save/load we have designed and built two crates:
+- [versionize](https://github.com/firecracker-microvm/versionize) - defines the Versionize trait, implements serialization of primitive types and provides a helper class to map Firecracker versions to individual structure versions.
+- [versionize_derive](https://github.com/firecracker-microvm/versionize_derive) - exports a procedural macro that consumes structures, enums or unions and their annotations to produce an implementation of the `Versionize` trait
```

Comment:
Should we link here the ones from crates.io instead?

Reply:
Done